File: packages/gaia-explorer/src/gaia_explorer/output.py
"""Helpers for building and persisting Gaia Explorer output payloads."""
from __future__ import annotations

import gzip
import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Dict

from .config import GridConfig, QueryConfig


def build_output(
    cubes: Dict[str, Dict[str, object]],
    grid: GridConfig,
    query: QueryConfig,
    retrieved: int,
    retained: int,
) -> Dict[str, object]:
    metadata = {
        "cube_edge_ly": grid.grid_size,
        "range_half_extent_ly": grid.half_extent,
        "query_radius_ly": query.radius_ly,
        "retrieved_rows": retrieved,
        "retained_stars": retained,
        "cubes_nonempty": len(cubes),
        "generated_utc": datetime.now(timezone.utc).isoformat(),
    }

    selection = {
        "parallax_min_mas": query.parallax_min_mas,
        "parallax_over_error_min": query.parallax_over_error_min,
        "ruwe_max": query.ruwe_max,
    }
    if query.temperature_min is not None:
        selection["temperature_min_K"] = query.temperature_min
    if query.limit is not None:
        selection["row_limit"] = query.limit

    return {
        "metadata": metadata,
        "selection": selection,
        "cubes": cubes,
    }


def dump_outputs(payload: Dict[str, object], json_path: Path, binary_path: Path) -> None:
    json_text = json.dumps(payload, separators=(",", ":"), ensure_ascii=False)
    json_path.write_text(json_text, encoding="utf-8")

    with gzip.open(binary_path, "wb") as handle:
        handle.write(json_text.encode("utf-8"))

File: packages/gaia-explorer/src/output.ts
import { writeFile, mkdir } from "node:fs/promises";
import { dirname, resolve } from "node:path";
import { promisify } from "node:util";
import { gzip } from "node:zlib";

import type { DeepReadonly } from "@cosmos-journeyer/typescript";

import type { GridConfig, QueryConfig } from "./config";
import { computeParallaxMinMas } from "./config";
import type { CubeId, SpatialCube } from "./spatial";
export type OutputPayload = DeepReadonly<{
    metadata: {
        cube_edge_ly: number;
        range_half_extent_ly: number;
        query_radius_ly: number;
        retrieved_rows: number;
        retained_stars: number;
        cubes_nonempty: number;
        generated_utc: string;
    };
    selection: {
        parallax_min_mas: number;
        parallax_over_error_min: number;
        ruwe_max: number;
        temperature_min_K?: number;
        row_limit?: number;
    };
    cubes: Readonly<Record<CubeId, SpatialCube>>;
}>;
export function buildOutput(
    cubes: ReadonlyMap<CubeId, SpatialCube>,
    grid: GridConfig,
    query: QueryConfig,
    retrieved: number,
    retained: number,
    now: () => Date = () => new Date(),
): OutputPayload {
    const selection: OutputPayload["selection"] = {
        parallax_min_mas: computeParallaxMinMas(query),
        parallax_over_error_min: query.parallaxOverErrorMin,
        ruwe_max: query.ruweMax,
        ...(query.temperatureMin === undefined ? {} : { temperature_min_K: query.temperatureMin }),
        ...(query.limit === undefined ? {} : { row_limit: query.limit }),
    };
    return {
        metadata: {
            cube_edge_ly: grid.gridSize,
            range_half_extent_ly: grid.halfExtent,
            query_radius_ly: query.radiusLy,
            retrieved_rows: retrieved,
            retained_stars: retained,
            cubes_nonempty: cubes.size,
            generated_utc: now().toISOString(),
        },
        selection,
        cubes: Object.fromEntries(cubes),
    };
}
export async function dumpOutputs(payload: OutputPayload, jsonPath: string, gzipPath: string): Promise<void> {
    const text = JSON.stringify(payload);
    await Promise.all([
        mkdir(dirname(resolve(jsonPath)), { recursive: true }),
        mkdir(dirname(resolve(gzipPath)), { recursive: true }),
    ]);
    await Promise.all([
        writeFile(jsonPath, text, "utf8"),
        promisify(gzip)(Buffer.from(text)).then(async (data) => writeFile(gzipPath, data)),
    ]);
}
